parquet_derive/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+  let fields: Fields = match input.data {
+    Data::Struct(DataStruct { fields, .. }) => fields,
+    Data::Enum(_) => unimplemented!("don't support enum"),
+    Data::Union(_) => unimplemented!("don't support union"),
+  };
+
+  let field_infos: Vec<FieldInfo> =
+    fields.iter().map(move |f| FieldInfo::from(f)).collect();
+
+  // Hand the output tokens back to the compiler
+  proc_macro::TokenStream::from(writer_from_field_infos(
+    input.ident,
+    input.generics,
+    &field_infos[..],
+  ))
+}
+
+fn extract_type_reference_info(
+  &syn::TypeReference {
+    ref lifetime,
+    ref elem,
+    ..
+  }: &syn::TypeReference,
+) -> (Ident, Option<Lifetime>, bool)
+{
+  if let Type::Path(ref type_path) = elem.as_ref() {
+    let (ident, is_option, _) = extract_path_info(type_path);
+    (ident, lifetime.clone(), is_option)
+  } else {
+    unimplemented!("unsupported elem {:#?}", elem)
+  }
+}
+
+fn extract_path_info(
+  &syn::TypePath {
+    path: Path { ref segments, .. },
+    ..
+  }: &syn::TypePath,
+) -> (Ident, bool, Option<PathArguments>)
+{
+  let seg = segments
+    .iter()
+    .next()
+    .expect("must have at least 1 segment");
+
+  let second_level = match &seg.arguments {
+    PathArguments::None => None,
+    PathArguments::AngleBracketed(AngleBracketedGenericArguments {
+      ref args, ..
+    }) => {
+      let generic_argument = args.iter().next().unwrap();
+      match generic_argument {
+        syn::GenericArgument::Lifetime(_) => unimplemented!("generic argument lifetime"),
+        syn::GenericArgument::Type(Type::Reference(ref tr)) => {
+          let (ident2, lifetime2, is_option2) = extract_type_reference_info(tr);
+          Some((ident2, lifetime2, is_option2))
+        },
+        syn::GenericArgument::Type(Type::Path(ref tp)) => {
+          let (ident2, _is_option2, _path_arguments) = extract_path_info(tp);
+          Some((ident2, None, false))
+        },
+        syn::GenericArgument::Type(_) => {
+          unimplemented!("generic argument type not reference/path")
+        },
+        syn::GenericArgument::Binding(_) => unimplemented!("generic argument binding"),
+        syn::GenericArgument::Constraint(_) => {
+          unimplemented!("generic argument constraint")
+        },
+        syn::GenericArgument::Const(_) => unimplemented!("generic argument const"),
+      }
+    },
+    PathArguments::Parenthesized(_) => unimplemented!("parenthesized"),
+  };
+
+  if &seg.ident.to_string()[..] == "Option" {
+    if let Some((ident2, _lifetime2, _is_option2)) = second_level {
+      (ident2, true, None)
+    } else {
+      unimplemented!("I couldn't parse what was inside of option")
+    }
+  } else {
+    (seg.ident.clone(), false, Some(seg.arguments.clone()))
+  }
+}
+
+impl FieldInfo {
+  fn from(f: &Field) -> Self {
+    let (field_type, field_lifetime, is_option) = match &f.ty {
+      Type::Slice(_) => unimplemented!("unsupported type: Slice"),
```

Comment:
My concern here is handling user-reported errors. I can do a generic fall-through but the error can't include what generated the error. Instead of this match it would be something like:

```
match &f.ty {
      Type::Reference(ref tr) => {
        //                unimplemented!("lifetime: {:#?}", lifetime);
        extract_type_reference_info(tr)
      },
      Type::Path(tp) => {
        let (ident, is_option, _) = extract_path_info(&tp);
        (ident, None, is_option)
      },
      _ => unimplemented!("unsupported type")
    }
```

so if someone tries to derive:

```
#[derive(ParquetRecordWriter)]
struct Foo<'a> {
  baz: String
  bar: &'a [u8]
}
```

it will just panic at compile time with `unsupported type` and I don't think the user will know right away which field caused the failure. This was a user-friendly consideration.

If we're OK omitting type-failure information then I'm OK simplifying, it will remove quite a few of lines.

Reply:
I just got to see this error in action when I tried compiling to test other fields into ByteArray:

```
not yet implemented: don't know Vec for Field
```

it's not the best error but it let's me know the Vec is the issue.

Reply:
@sunchao what do you think about this response?

Reply:
@xrl sorry I just saw your comments. In the example you gave above, maybe we can do this:
```rust
match &f.ty {
      Type::Reference(ref tr) => {
        //                unimplemented!("lifetime: {:#?}", lifetime);
        extract_type_reference_info(tr)
      },
      Type::Path(tp) => {
        let (ident, is_option, _) = extract_path_info(&tp);
        (ident, None, is_option)
      },
      t @ _ => unimplemented!("unsupported type {}", t)
    }
```
assuming `Type` implements `Display` or `Debug`. What do you think?

Reply:
This is great, thanks!